gen-kubectldocs/generators/html.go

```diff
@@ -0,0 +1,371 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package generators
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+)
+
+type TOCItem struct {
+	Level       int
+	Title       string
+	Link        string
+	Buffer      string
+	SubSections []*TOCItem
+}
+
+type TOC struct {
+	Title     string
+	Copyright string
+	Sections  []*TOCItem
+}
+
+type HTMLWriter struct {
+	TOC            TOC
+	CurrentSection *TOCItem
+}
+
+func NewHTMLWriter(copyright, title string) DocWriter {
+	writer := HTMLWriter{
+		TOC: TOC{
+			Copyright: copyright,
+			Title:     title,
+			Sections:  []*TOCItem{},
+		},
+	}
+	return &writer
+}
+
+func (h *HTMLWriter) Extension() string {
+	return ".html"
+}
+
+func (h *HTMLWriter) WriteCommandFiles(toc ToC, params KubectlSpec) {
+	/* map the top level command name to command structure */
+	m := map[string]TopLevelCommand{}
+	for _, g := range params.TopLevelCommandGroups {
+		for _, tlc := range g.Commands {
+			m[tlc.MainCommand.Name] = tlc
+		}
+	}
+
+	// REVISIT, use the toc.yaml file to map the tlc to the category/group
+	for _, c := range toc.Categories {
+
+		// Write each of the commands in this category
+		for _, cm := range c.Commands {
```

Comment:
any consideration on the ordering of values in the map?

Reply:
maybe it is okay at this stage ... eventually we need an ordered list of the commands and options so that the output can be easily compared to each other.

Reply:
Yes, thinking of sorting the commands based on the order/grouping of commands in toc.yaml. I updated toc.yaml to match the output of `kubectl help`.
Ideally, the group name/message would be provided by the spec (or kubectl help).

Reply:
The commands are ordered per category. This is fixed in toc.yaml.
I guess the order of the sub-commands and options could change from release to release. The rate of change seems kind of slow, though.

Reply:
ok